File: Python/three_sum_closest.py
# https://leetcode.com/problems/3sum-closest/description/
from utils.examination import Examiner, Testcase, Logger
from utils.search import Searcher, SearchResult, SearchQuery, Range, Element

class Solution:
    def threeSumClosest(self, nums: list[int], target: int) -> int:
        nums.sort()

        if 0 < nums[0] and target <= nums[0]:
            return nums[0] + nums[1] + nums[2]
        if 0 > nums[-1] and target <= nums[-1]:
            return nums[-1] + nums[-2] + nums[-3]
        searcher: Searcher = Searcher.in_(nums)
        min_diff = float('+inf')
        closest_sum_: int = 0
        zero: SearchResult = searcher.search(SearchQuery(0))
        if zero.has_found():
            pointers: tuple[int, int] = (zero.index, zero.index + 1)
        else:
            first_pointer: Element = zero.nearest_left
            second_pointer: Element = zero.nearest_right
            if first_pointer.is_out_of_range():
                first_pointer = Element(nums[0], 0)
                second_pointer = Element(nums[1], 1)
            if second_pointer.is_out_of_range():
                first_pointer = Element(nums[-2], len(nums) - 2)
                second_pointer = Element(nums[-1], len(nums) - 1)
            pointers: tuple[int, int] = (first_pointer.index, second_pointer.index)
        while 0 <= pointers[0] and len(nums) - 1 >= pointers[1]:
            remaining_value: int = target - (nums[pointers[0]] + nums[pointers[1]])
            if 0 <= remaining_value:
                range_: Range = Range.start_from(pointers[1] + 1)
            else:
                range_: Range = Range.till(pointers[0] - 1)
            if 0 == pointers[0] and len(nums) - 1 == pointers[1]:
                range_: Range = Range(1, len(nums) - 2)
            remaining_search_result: SearchResult = searcher.search(SearchQuery(remaining_value, range_))
            if remaining_search_result.has_found():
                return target
            remaining: Element = remaining_search_result.nearest_to(remaining_value)

            sum_: int = remaining.value + (nums[pointers[0]] + nums[pointers[1]])
            diff: int = abs(target - sum_)
            Logger.log(min_diff, diff, pointers[0], pointers[1], remaining.value, remaining_search_result, remaining_value, range_)
            if min_diff > diff:
                min_diff = diff
                closest_sum_ = sum_

            if 0 == pointers[0] and len(nums) - 1 == pointers[1]:
                break
            if target <= sum_:
                s: int = pointers[0] - 1
                if 0 > s:
                    pointers: tuple[int, int] = (0, pointers[1] + 1)
                    continue
                pointers: tuple[int, int] = (s, pointers[1])
            else:
                e: int = pointers[1] + 1
                if len(nums) - 1 < e:
                    pointers: tuple[int, int] = (pointers[0] - 1, len(nums) - 1)
                    continue
                pointers: tuple[int, int] = (pointers[0], e)
        return closest_sum_

def examine()->None:
    Examiner(Solution(), 'threeSumClosest').run((
        Testcase(([1,1,-1], 2), -1),
        Testcase(([1,1,1,1], 3), 3),
        Testcase(([1,1,1,0], -100), 2),
        Testcase(([-100,-98,-2,-1], -101), -101),
        Testcase(([-1,2,1,-4], 1), 2),
        Testcase(([-100, -2, 1, 3, 25, 75], 0), 0),
        Testcase(([102, 4, 3, -23, -73], 2), 6),
        Testcase(([102, 4, 3, 2, 2, 2, -23, -73], 2), 6),
    ))

File: TypeScript/solutions/three-sum-closest.ts
// https://leetcode.com/problems/3sum-closest/description/
import { Examiner, Testcase, Searcher, SearchQuery, Range, Element } from '../utils';


function threeSumClosest(nums: number[], target: number): number {
    nums.sort((a, b) => a - b);

    if (nums[0]! > 0 && target <= nums[0]!) return nums[0]! + nums[1]! + nums[2]!;
    if (nums[nums.length - 1]! < 0 && target <= nums[nums.length - 1]!) {
        return nums[nums.length - 1]! + nums[nums.length - 2]! + nums[nums.length - 3]!;
    }

    const searcher = Searcher.in_(nums);
    let minDiff = Infinity;
    let closestSum = 0;

    const zero = searcher.search(new SearchQuery(0));
    let pointers: [number, number];
    if (zero.hasFound()) {
        pointers = [zero.index, zero.index + 1];
    } else {
        let firstPointer = zero.nearestLeft;
        let secondPointer = zero.nearestRight;
        if (firstPointer.isOutOfRange()) {
            firstPointer = new Element(nums[0]!, 0);
            secondPointer = new Element(nums[1]!, 1);
        }
        if (secondPointer.isOutOfRange()) {
            firstPointer = new Element(nums[nums.length - 2]!, nums.length - 2);
            secondPointer = new Element(nums[nums.length - 1]!, nums.length - 1);
        }
        pointers = [firstPointer.index!, secondPointer.index!];
    }

    while (pointers[0] >= 0 && pointers[1] <= nums.length - 1) {
        const remainingValue = target - (nums[pointers[0]]! + nums[pointers[1]]!);
        let range: Range;
        if (remainingValue >= 0) {
            range = Range.startFrom(pointers[1] + 1);
        } else {
            range = Range.till(pointers[0] - 1);
        }
        if (pointers[0] === 0 && pointers[1] === nums.length - 1) {
            range = new Range(1, nums.length - 2);
        }

        const remainingResult = searcher.search(new SearchQuery(remainingValue, range));
        if (remainingResult.hasFound()) return target;
        const remaining = remainingResult.nearestTo(remainingValue);

        const sum = remaining.value! + (nums[pointers[0]]! + nums[pointers[1]]!);
        const diff = Math.abs(target - sum);
        if (minDiff > diff) { minDiff = diff; closestSum = sum; }

        if (pointers[0] === 0 && pointers[1] === nums.length - 1) break;
        if (target <= sum) {
            const s = pointers[0] - 1;
            if (s < 0) { pointers = [0, pointers[1] + 1]; continue; }
            pointers = [s, pointers[1]];
        } else {
            const e = pointers[1] + 1;
            if (e > nums.length - 1) { pointers = [pointers[0] - 1, nums.length - 1]; continue; }
            pointers = [pointers[0], e];
        }
    }
    return closestSum;
}

export function examine(): void {
    new Examiner(threeSumClosest).run([
        new Testcase([[1, 1, -1], 2], 1),
        new Testcase([[1, 1, 1, 1], 3], 3),
        new Testcase([[1, 1, 1, 0], -100], 2),
        new Testcase([[-100, -98, -2, -1], -101], -101),
        new Testcase([[-1, 2, 1, -4], 1], 2),
        new Testcase([[-100, -2, 1, 3, 25, 75], 0], 0),
        new Testcase([[102, 4, 3, -23, -73], 2], 6),
        new Testcase([[102, 4, 3, 2, 2, 2, -23, -73], 2], 6),
    ]);
}
